File: android/app/src/main/kotlin/ai/ownward/app/data/ImageEncoder.kt
package ai.ownward.app.data

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.Base64
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream

// 服务端约束（src/chat-images.ts）：png/jpeg/webp/gif，≤6张/条，≤5MB/张（解码后），
// ≤12MB/条，裸 base64 且魔数必须匹配声明类型。HEIC 等一律本地转码成 JPEG。
object ImageEncoder {
    const val MAX_IMAGES = 6
    private const val MAX_BYTES = 5 * 1024 * 1024
    private const val SCALE_TARGET = 1920

    /** 返回 null 表示读取失败 */
    suspend fun fromUri(context: Context, uri: Uri): OutImage? = withContext(Dispatchers.IO) {
        val resolver = context.contentResolver
        val declared = resolver.getType(uri) ?: ""
        val raw = resolver.openInputStream(uri)?.use { it.readBytes() } ?: return@withContext null

        // gif 保持原样（重编码会丢动画）；超限的 gif 直接拒绝，由调用方提示
        if (declared == "image/gif" && sniff(raw) == "image/gif") {
            return@withContext if (raw.size <= MAX_BYTES)
                OutImage("image/gif", Base64.encodeToString(raw, Base64.NO_WRAP)) else null
        }

        // png/jpeg/webp 且不超限：原样发，媒体类型以魔数为准（服务端会严格校验）
        val sniffed = sniff(raw)
        if (sniffed != null && raw.size <= MAX_BYTES) {
            return@withContext OutImage(sniffed, Base64.encodeToString(raw, Base64.NO_WRAP))
        }

        // 其余（HEIC、超大图）：解码位图重编码为 JPEG，必要时缩边
        var bmp = BitmapFactory.decodeByteArray(raw, 0, raw.size) ?: return@withContext null
        val longest = maxOf(bmp.width, bmp.height)
        if (longest > SCALE_TARGET) {
            val scale = SCALE_TARGET.toFloat() / longest
            bmp = Bitmap.createScaledBitmap(
                bmp, (bmp.width * scale).toInt(), (bmp.height * scale).toInt(), true
            )
        }
        var quality = 88
        var out: ByteArray
        do {
            val buf = ByteArrayOutputStream()
            bmp.compress(Bitmap.CompressFormat.JPEG, quality, buf)
            out = buf.toByteArray()
            quality -= 12
        } while (out.size > MAX_BYTES && quality > 30)
        if (out.size > MAX_BYTES) return@withContext null
        OutImage("image/jpeg", Base64.encodeToString(out, Base64.NO_WRAP))
    }

    /** 服务端按魔数校验，所以这里也按魔数判型而不是信 ContentResolver */
    private fun sniff(b: ByteArray): String? = when {
        b.size > 8 && b[0] == 0x89.toByte() && b[1] == 0x50.toByte() -> "image/png"
        b.size > 3 && b[0] == 0xFF.toByte() && b[1] == 0xD8.toByte() -> "image/jpeg"
        b.size > 12 && b[0] == 'R'.code.toByte() && b[8] == 'W'.code.toByte() && b[9] == 'E'.code.toByte() -> "image/webp"
        b.size > 6 && b[0] == 'G'.code.toByte() && b[1] == 'I'.code.toByte() && b[2] == 'F'.code.toByte() -> "image/gif"
        else -> null
    }
}

File: ios/Ownward/Core/ImageEncoder.swift
// 服务端约束（src/chat-images.ts）：png/jpeg/webp/gif，≤6张/条，≤5MB/张，≤12MB/条，
// 裸 base64 且魔数必须匹配声明类型。HEIC 等一律本地转码成 JPEG（对齐 android ImageEncoder.kt）。
import UIKit

enum ImageEncoder {
    static let maxImages = 6
    private static let maxBytes = 5 * 1024 * 1024
    private static let scaleTarget: CGFloat = 1920

    /// 返回 nil 表示无法编码（读取失败或压到 30% 仍超 5MB）
    static func encode(_ raw: Data) -> OutImage? {
        let sniffed = sniff(raw)
        // gif 保持原样（重编码会丢动画）；超限的 gif 直接拒绝
        if sniffed == "image/gif" {
            return raw.count <= maxBytes ? OutImage(media_type: "image/gif", data: raw.base64EncodedString()) : nil
        }
        if let sniffed, raw.count <= maxBytes {
            return OutImage(media_type: sniffed, data: raw.base64EncodedString())
        }
        guard var image = UIImage(data: raw) else { return nil }
        let longest = max(image.size.width, image.size.height) * image.scale
        if longest > scaleTarget {
            let scale = scaleTarget / longest
            let size = CGSize(width: image.size.width * image.scale * scale, height: image.size.height * image.scale * scale)
            let fmt = UIGraphicsImageRendererFormat()
            fmt.scale = 1
            image = UIGraphicsImageRenderer(size: size, format: fmt).image { _ in
                image.draw(in: CGRect(origin: .zero, size: size))
            }
        }
        var quality: CGFloat = 0.88
        while true {
            guard let out = image.jpegData(compressionQuality: quality) else { return nil }
            if out.count <= maxBytes { return OutImage(media_type: "image/jpeg", data: out.base64EncodedString()) }
            quality -= 0.12
            if quality <= 0.30 { return nil }
        }
    }

    /// 服务端按魔数校验，所以这里也按魔数判型
    static func sniff(_ b: Data) -> String? {
        let bytes = [UInt8](b.prefix(16))
        if bytes.count > 8, bytes[0] == 0x89, bytes[1] == 0x50 { return "image/png" }
        if bytes.count > 3, bytes[0] == 0xFF, bytes[1] == 0xD8 { return "image/jpeg" }
        if bytes.count > 12, bytes[0] == 0x52, bytes[8] == 0x57, bytes[9] == 0x45 { return "image/webp" }
        if bytes.count > 6, bytes[0] == 0x47, bytes[1] == 0x49, bytes[2] == 0x46 { return "image/gif" }
        return nil
    }
}
